dev-tools/smoke.js

```diff
@@ -0,0 +1,435 @@
+		// Deliberately don't print a `\n` character so that we can overwrite
+		process.stdout.write(
+			`[${this.format}] P:${p.pokemon} I:${p.items} A:${p.abilities} M:${p.moves} = ${this.games}`);
+	}
+}
+
+module.exports = {TeamGenerator, Pool, CoordinatedPlayerAI, SmokeRunner};
+
+function onlyValid(gen, obj, getter, additional, nonStandard) {
+	return Object.keys(obj).filter(k => {
+		const v = getter(k);
+		return v.gen <= gen &&
+			(!v.isNonstandard || !!nonStandard) &&
+			(!additional || additional(k, v));
+	});
+}
+
+function getSignatures(dex, pools) {
+	const signatures = new Map();
+	for (const id of pools.items.possible) {
+		const item = dex.data.Items[id];
+		if (item.megaEvolves) {
+			const pokemon = toId(item.megaEvolves);
+			const combo = {item: id};
+			let combos = signatures.get(pokemon);
+			if (!combos) {
+				combos = [];
+				signatures.set(pokemon, combos);
+			}
+			combos.push(combo);
+		} else if (item.zMoveUser) {
+			for (const user of item.zMoveUser) {
+				const pokemon = toId(user);
+				const combo = {item: id};
+				if (item.zMoveFrom) combo.move = toId(item.zMoveFrom);
+				let combos = signatures.get(pokemon);
+				if (!combos) {
+					combos = [];
+					signatures.set(pokemon, combos);
+				}
+				combos.push(combo);
+			}
+		}
+	}
+	return signatures;
+}
+
+if (require.main === module) {
+	Dex.includeModData();
+	RejectionTracker.register();
+
+	// TODO: Add triple battles once supported by the AI.
+	const FORMATS = [
+		'gen7customgame', 'gen7doublescustomgame',
+		'gen6customgame', 'gen6doublescustomgame',
+		'gen5customgame', 'gen5doublescustomgame',
+		'gen4customgame', 'gen4doublescustomgame',
+		'gen3customgame', 'gen3doublescustomgame',
+		'gen2customgame',
+		'gen1customgame'];
+
+	// Because of our handwavy accounting of 'usage', as well as to account for ordering effects and
+	// the interplay of various combinations, we should actually run multiple cycles and exhaust our
+	// pools multiple times to provide more confidence that we're actually smoking out issues.
+	//
+	// If cycles is negative the smoke test will instead run indefinitely in case you wish to
+	// dedicate some computer resources to finding PS! edge case crashes instead of mining crypto,
+	// folding proteins, or searching for extraterrestrials/large primes.
+	const cycles = Number(process.argv[2]) || DEFAULT_CYCLES;
+	const format = process.argv[3];
+	const formats = format ? [format] : FORMATS;
+	const maxFailures = format ? 1 : MAX_FAILURES;
+	const seed = process.argv[4] && process.argv[4].split(',').map(s => Number(s));
+	const prng = new PRNG(seed);
+	(async () => {
+		let failures = 0;
+		do {
+			for (let format of formats) {
+				failures += await new SmokeRunner({
+					format, cycles, prng, maxFailures, log: true,
+				}).run();
+				process.stdout.write('\n');
+				if (failures >= maxFailures) break;
+			}
+		} while (cycles < 0); // eslint-disable-line no-unmodified-loop-condition
```

Comment:
`repeatForever = cycles < 0; do ... while (repeatForever)`

is probably enough to shut up the linter, and is probably also more readable.

Reply:
Wait, no it's not.

Reply:
Still, it might be the best way to write this code.

```js
if (cycles < 0) {
  repeatForever = true;
  cycles = -cycles;
}
```

Reply:
I tried:

```js
const forever = cycles < 0;
do {
 ...
} while (forever);
```

And I still get:

```
/dev-tools/smoke.js
  429:12  error  'forever' is not modified in this loop  no-unmodified-loop-condition
```

Maybe the `forever` variable makes things more readable though? Iunno, I'm not convinced, because then it looks like the loop might always be an infinite loop at first glance.

Reply:
I think the code I suggest makes things more readable, yes, even after realizing it wouldn't shut up the linter, which I realized when I said "Wait, no it's not".